### DISCUSSION TITLE
Is there any proposals or specification that use spread operator within collection initializer?

### DISCUSSION THREAD

**Discussion by JerryJian** [dotnet/csharplang#8180] [Language Ideas]
Just like what was asked in the title, is there any proposals or specification that use spread operator within collection initializer?
Code example:
``` c#
class Options
{
    public List<string> SelectedIds { get; } = [];
}

class ViewModel
{
    public List<Checkable> CheckList { get; } = new();

    public async Task DoWork()
    {
        var options = new Options
        {
            SelectedIds =
            {
                ..CheckList.Where(t => t.IsChecked).Select(t => t.Id),
            }
        };
    }
}
```

Thanks!

**Comment by huoyaoyuan** [comment]
Unlike the old collection initializer, collection initializer doesn't work with the `Add` method on collections. So support for additions into existing collection is new.

For this proposal, there is an alternative approach with `List<string> SelectedIds { get; init; } = []`.

**Reply by JerryJian** [reply]
The code in the description is a hand writing example. The most common one I encounter is [`ProcessStartInfo.ArgumentList`](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.processstartinfo.argumentlist?view=net-8.0), which doesn't has setter.

> Unlike the old collection initializer, collection initializer doesn't work with the Add method on collections. So support for additions into existing collection is new.

I'm aware of this just now, I will study it.

**Reply by JerryJian** [reply]
> Unlike the old collection initializer, collection initializer doesn't work with the Add method on collections. So support for additions into existing collection is new.

But I use ILSpy to view the IL code like this:
```c#
public class InstallerXSite
{
	public InstallerXSite()
	{
		new ProcessStartInfo
		{
			ArgumentList = { "1", "2" }
		};
	}
}
```

```il
.class public auto ansi beforefieldinit InstallerXLib.InstallerXSite
	extends [System.Runtime]System.Object
{
	// Methods
	.method public hidebysig specialname rtspecialname 
		instance void .ctor () cil managed 
	{
		// Method begins at RVA 0x2050
		// Header size: 1
		// Code size: 47 (0x2f)
		.maxstack 8

		IL_0000: ldarg.0
		IL_0001: call instance void [System.Runtime]System.Object::.ctor()
		IL_0006: nop
		IL_0007: nop
		IL_0008: newobj instance void [System.Diagnostics.Process]System.Diagnostics.ProcessStartInfo::.ctor()
		IL_000d: dup
		IL_000e: callvirt instance class [System.Runtime]System.Collections.ObjectModel.Collection`1<string> [System.Diagnostics.Process]System.Diagnostics.ProcessStartInfo::get_ArgumentList()
		IL_0013: ldstr "1"
		IL_0018: callvirt instance void class [System.Runtime]System.Collections.ObjectModel.Collection`1<string>::Add(!0)
		IL_001d: nop
		IL_001e: callvirt instance class [System.Runtime]System.Collections.ObjectModel.Collection`1<string> [System.Diagnostics.Process]System.Diagnostics.ProcessStartInfo::get_ArgumentList()
		IL_0023: ldstr "2"
		IL_0028: callvirt instance void class [System.Runtime]System.Collections.ObjectModel.Collection`1<string>::Add(!0)
		IL_002d: nop
		IL_002e: ret
	} // end of method InstallerXSite::.ctor

} // end of class InstallerXLib.InstallerXSite
```

We can see it calls `Add` method to add the list one by one.

**Reply by huoyaoyuan** [reply]
Sorry for the typo. I was to mean that *collection expression doesn't work with `Add` method*.

**Reply by JerryJian** [reply]
Ok, thanks a lot. Finnaly I found that I could write an `Add(IEnumerable<T> collections)` extension method to achieve this, but the name `Add` may cause missunderstanding.
